File: labs/05-memory-system/go/main.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"strings"

	openaiclient "github.com/ZYZ-Labs/agent-labs/shared/config"
)

type shortTermMemory struct {
	messages    []openaiclient.Message
	maxMessages int
}

func newShortTermMemory(maxMessages int) *shortTermMemory {
	return &shortTermMemory{maxMessages: maxMessages}
}

func (m *shortTermMemory) add(role, content string) {
	m.messages = append(m.messages, openaiclient.Message{Role: role, Content: content})
	if len(m.messages) > m.maxMessages {
		keep := 0
		if len(m.messages) > 0 && m.messages[0].Role == "system" {
			keep = 1
		}
		m.messages = append(m.messages[:keep], m.messages[keep+1:]...)
	}
}

func (m *shortTermMemory) get() []openaiclient.Message {
	out := make([]openaiclient.Message, len(m.messages))
	copy(out, m.messages)
	return out
}

type longTermMemory struct {
	client *openaiclient.Client
}

func newLongTermMemory(client *openaiclient.Client) *longTermMemory {
	return &longTermMemory{client: client}
}

func (ltm *longTermMemory) embed(text string) []float64 {
	if ltm.client != nil {
		resp, err := ltm.client.ChatCompletion(openaiclient.ChatCompletionRequest{
			Messages: []openaiclient.Message{
				{Role: "user", Content: "Summarize in one sentence for retrieval: " + text},
			},
			MaxTokens: intPtr(20),
		})
		if err == nil && len(resp.Choices) > 0 {
			text = resp.Choices[0].Message.Content
		}
	}
	vec := make([]float64, 64)
	h := sha256.Sum256([]byte(text))
	for i, b := range h {
		vec[i%64] += float64(b) / 255.0
	}
	return vec
}

func (ltm *longTermMemory) store(text string, metadata map[string]string) {
	fmt.Println("[Long-term memory] Chroma not available, skipping store.")
}

func (ltm *longTermMemory) retrieve(query string, n int) []string {
	return nil
}

type agentWithMemory struct {
	client    *openaiclient.Client
	shortTerm *shortTermMemory
	longTerm  *longTermMemory
}

func newAgentWithMemory(client *openaiclient.Client) *agentWithMemory {
	return &agentWithMemory{
		client:    client,
		shortTerm: newShortTermMemory(12),
		longTerm:  newLongTermMemory(client),
	}
}

func (a *agentWithMemory) chat(userInput string) string {
	relevant := a.longTerm.retrieve(userInput)
	if len(relevant) > 0 {
		context := "Relevant memory:\n"
		for _, r := range relevant {
			context += "- " + r + "\n"
		}
		a.shortTerm.add("system", strings.TrimSpace(context))
	}

	a.shortTerm.add("user", userInput)
	resp, err := a.client.ChatCompletion(openaiclient.ChatCompletionRequest{
		Messages:  a.shortTerm.get(),
		MaxTokens: intPtr(200),
	})
	if err != nil {
		return fmt.Sprintf("Error: %v", err)
	}
	answer := resp.Choices[0].Message.Content
	a.shortTerm.add("assistant", answer)
	a.longTerm.store(fmt.Sprintf("User: %s\nAssistant: %s", userInput, answer), nil)
	return answer
}

func intPtr(i int) *int {
	return &i
}

func main() {
	client := openaiclient.NewClient()
	agent := newAgentWithMemory(client)

	fmt.Println("Agent:", agent.chat("My name is Alice and I work on backend systems."))
	fmt.Println("Agent:", agent.chat("What do I work on?"))
	fmt.Println("Agent:", agent.chat("Suggest a logging strategy for my team."))
}

File: labs/05-memory-system/java/Main.java
import com.agentlabs.OpenAiClient;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Main {

    public static void main(String[] args) throws Exception {
        OpenAiClient client = new OpenAiClient();
        AgentWithMemory agent = new AgentWithMemory(client);

        System.out.println("Agent: " + agent.chat("My name is Alice and I work on backend systems."));
        System.out.println("Agent: " + agent.chat("What do I work on?"));
        System.out.println("Agent: " + agent.chat("Suggest a logging strategy for my team."));
    }

    static class ShortTermMemory {
        private final List<ObjectNode> messages = new ArrayList<>();
        private final int maxMessages;

        ShortTermMemory(int maxMessages) {
            this.maxMessages = maxMessages;
        }

        void add(String role, String content) {
            ObjectMapper mapper = new ObjectMapper();
            ObjectNode msg = mapper.createObjectNode();
            msg.put("role", role);
            msg.put("content", content);
            messages.add(msg);
            if (messages.size() > maxMessages) {
                int keep = (!messages.isEmpty() && "system".equals(messages.get(0).get("role").asText())) ? 1 : 0;
                messages.subList(keep, keep + 1).clear();
            }
        }

        ArrayNode get() {
            ObjectMapper mapper = new ObjectMapper();
            ArrayNode array = mapper.createArrayNode();
            messages.forEach(array::add);
            return array;
        }
    }

    static class LongTermMemory {
        private final List<String> entries = new ArrayList<>();

        void store(String text) {
            entries.add(text);
        }

        List<String> retrieve(String query) {
            String[] queryWords = query.toLowerCase().split("\\s+");
            return entries.stream()
                    .sorted((a, b) -> Integer.compare(score(b, queryWords), score(a, queryWords)))
                    .limit(3)
                    .filter(e -> score(e, queryWords) > 0)
                    .toList();
        }

        private int score(String entry, String[] queryWords) {
            String lower = entry.toLowerCase();
            return (int) Arrays.stream(queryWords).filter(lower::contains).count();
        }
    }

    static class AgentWithMemory {
        private final OpenAiClient client;
        private final ShortTermMemory shortTerm;
        private final LongTermMemory longTerm;

        AgentWithMemory(OpenAiClient client) {
            this.client = client;
            this.shortTerm = new ShortTermMemory(12);
            this.longTerm = new LongTermMemory();
        }

        String chat(String userInput) throws Exception {
            List<String> relevant = longTerm.retrieve(userInput);
            if (!relevant.isEmpty()) {
                String context = "Relevant memory:\n" + String.join("\n", relevant.stream().map(r -> "- " + r).toList());
                shortTerm.add("system", context);
            }

            shortTerm.add("user", userInput);
            ObjectMapper mapper = new ObjectMapper();
            ObjectNode payload = mapper.createObjectNode();
            payload.set("messages", shortTerm.get());
            payload.put("max_tokens", 200);

            JsonNode response = client.chatCompletion(payload);
            String answer = client.extractMessage(response).get("content").asText();
            shortTerm.add("assistant", answer);
            longTerm.store("User: " + userInput + "\nAssistant: " + answer);
            return answer;
        }
    }
}
